firmware/modes/programmer/at89/at89.c

```diff
@@ -0,0 +1,654 @@
+#include "at89.h"
+
+static zif_bits_t zbits_null = {0, 0, 0, 0, 0};
+static zif_bits_t gnd        = {0, 0, 8, 0, 0};
+static zif_bits_t vdd        = {0, 0, 4, 0, 128};
+static zif_bits_t vpp        = {0, 0, 0, 64, 0};
+
+inline void print_banner(void)
+{
+    com_println("   | |");
+    com_println(" ==[+]==  open-tl866 Programmer Mode (AT89)");
+    com_println("   | |");
+}
+inline void print_help(void)
+{
+    com_println("\r\nCommands:\r\n  r <ADDR (hex)> [RANGE (hex)]\tRead from target");
+    com_println("  w <ADDR (hex)> <BYTE (hex)>\tWrite to target");
+    com_println("  e\t\t\t\tErase target");
+    com_println("  l <MODE (int)>\t\tSet lock bits to MODE");
+    com_println("  s\t\t\t\tPrint signature bytes");
+    com_println("  b\t\t\t\tBlank check");
+    com_println("  T\t\t\t\tRun some tests");
+    com_println("  h\t\t\t\tPrint help\r\n  v\t\t\t\tPrint version(s)");
+}
+
+inline void print_version()
+{
+    // All these should be defined in some config header files. TODO
+    com_println("Programmer Mode - AT89 version: 0.0.1");
+    com_println("open-tl866 lib version: UNIMPLEMENTED");
+    com_println("");
+}
+
+// Neat trick taken from a stack overflow answer.
+inline unsigned char invert_bit_endianness(unsigned char byte)
+{
+    static unsigned char lookup[16] = {
+                            0x0, 0x8, 0x4, 0xc, 0x2, 0xa, 0x6, 0xe,
+                            0x1, 0x9, 0x5, 0xd, 0x3, 0xb, 0x7, 0xf, };
+    return (lookup[byte & 0b1111] << 4) | lookup[byte >> 4];
+}
+
+inline void mask_p2_7(zif_bits_t op_base)
+{
+    op_base[3] |= 0x8;
+}
+
+inline void mask_p3_6(zif_bits_t op_base)
+{
+    op_base[1] |= 0x80;
+}
+
+inline void mask_p3_7(zif_bits_t op_base)
+{
+    op_base[2] |= 0x1;
+}
+
+inline void mask_xtal1(zif_bits_t op_base)
+{
+    op_base[2] |= 0x4;
+}
+
+inline void mask_prog(zif_bits_t op_base)
+{
+    op_base[3] |= 0x20;
+}
+
+inline void mask_addr(zif_bits_t op_base, unsigned int addr)
+{
+    op_base[0] = addr & 0xFF;
+    op_base[2] |= (addr >> 8) << 4;
+}
+
+inline void mask_data(zif_bits_t op_base, unsigned char data)
+{
+    op_base[3] |= (data & 0x80);
+    op_base[4] |= (invert_bit_endianness(data & 0x7f) >> 1);
+}
+
+inline unsigned char zif_to_data(zif_bits_t zif_state)
+{
+    // Filter the zif_bits response into a char byte with P0 bits
+                   // Trim non-data ZIF pins    // Set the LSB of data byte
+    unsigned char byte = (zif_state[4] << 1) | !! (zif_state[3] & (1 << 7));
+
+    // Invert bit-endianness
+    return invert_bit_endianness(byte);

```

Comment:
why is this needed?

Reply:
This is used by the read_byte(..) and read_sig() functions. Pins P0.0 through P0.7 are set to represent the requested byte, so this converts it to an unsigned char for later use.